packages/hub/src/session.ts

```diff
@@ -0,0 +1,123 @@
+import { Session as SessionInterface, SessionContext, SessionStatus } from '@sentry/types';
+import { dropUndefinedKeys, uuid4 } from '@sentry/utils';
+
+/**
+ * @inheritdoc
+ */
+export class Session implements SessionInterface {
+  public userAgent?: string;
+  public errors: number = 0;
+  public release?: string;
+  public sid: string = uuid4();
+  public did?: string;
+  public timestamp: number = Date.now();
+  public started: number = Date.now();
+  public duration: number = 0;
+  public status: SessionStatus = SessionStatus.Ok;
+  public environment?: string;
+  public ipAddress?: string;
+
+  constructor(context?: Omit<SessionContext, 'started' | 'status'>) {
+    if (context) {
+      this.update(context);
+    }
+  }
+
+  /** JSDoc */
+  // eslint-disable-next-line complexity
+  update(context: SessionContext = {}): void {
+    if (context.user) {
+      if (context.user.ip_address) {
+        this.ipAddress = context.user.ip_address;
+      }
+
+      if (!context.did) {
+        this.did = context.user.id || context.user.email || context.user.username;
+      }
+    }
+
+    this.timestamp = context.timestamp || Date.now();
+
+    if (context.sid) {
+      // Good enough uuid validation. — Kamil
+      this.sid = context.sid.length === 32 ? context.sid : uuid4();
+    }
+    if (context.did) {
+      this.did = `${context.did}`;
+    }
+    if (typeof context.started === 'number') {
+      this.started = context.started;
+    }
+    // I'm not sure why we even allow to set that, while having `started` and `timestamp` already. — Kamil
+    if (typeof context.duration === 'number') {
+      this.duration = context.duration;
+    } else {
+      this.duration = this.timestamp - this.started;
+    }
+    if (context.release) {
+      this.release = context.release;
+    }
+    if (context.environment) {
+      this.environment = context.environment;
+    }
+    if (context.ipAddress) {
+      this.ipAddress = context.ipAddress;
+    }
+    if (context.userAgent) {
+      this.userAgent = context.userAgent;
+    }
+    if (typeof context.errors === 'number') {
+      this.errors = context.errors;
+    }
+    if (context.status) {
+      this.status = context.status;
+    }
+  }
+
+  /** JSDoc */
+  close(status?: SessionStatus): void {
+    if (status) {
+      this.update({ status });
+    } else if (this.status === SessionStatus.Ok) {
+      this.update({ status: SessionStatus.Exited });
```

Comment:
This is weird: `session.close(Session.Status.Ok)` and `session.close()` produce different results, respectively `status == Ok` and `status == Exited`.

Reply:
Removed one of the unnecessary branches. (doesnt fully fix the case you described, but I'd say it's an edge case not work being defensive against).

Reply:
This is missing now the `update` call (to update timestamps) if neither conditions above are met.

I think calling `close` should only accept [terminal session states](https://develop.sentry.dev/sdk/sessions/#terminal-session-states), therefore `SessionStatus.Ok` is not a valid argument.

```js
close(status?: TerminalSessionStatus): void {
  if (!status) {
    // status = one of Exited | Crashed | Abnormal, depending on current state of Session;
  }
  this.update({ status });
}
```

Reply:
By calling `hub.endSession()`, you don't know what's sessions state, so we have to check whether it was in `Ok` state to turn it into `Exited`. I reverted it to the initial solution and updated the types to only use terminal states.